Changes to gem5 Resources [gem5 Vision]

kunpai opened the pull request:
We added a config file to configs/
We changes resources in src/python/gem5/resources by adding an api folder with mongoclient and jsonclient
We also made corresponding changes to testing directory and resource python files to make it work with our new schema

kunpai committed:
Changed tests to what work with our implementation

Co-authored-by: Parth Shah <helloparthshah@gmail.com>
Co-authored-by: Harshil Patel <harshilp2107@gmail.com>

helloparthshah committed:
Wrote some tests

Co-authored-by: Harshil2107 <harshilp2107@gmail.com>
Co-authored-by: kunpai <kunpai@ucdavis.edu>

kunpai committed:
Merge branch 'darchr:develop' into gem5Vision

Harshil2107 committed:
fixed typos

helloparthshah committed:
Wrote mongoclient and jsonclient

Co-authored-by: kunpai <kunpai@ucdavis.edu>
Co-authored-by: Harshil2107 <harshilp2107@gmail.com>

Harshil2107 committed:
cleaned up api code

Harshil2107 committed:
edited test for api

helloparthshah committed:
Created abstract client for json and mongodb
Created wrapper for that creates clients from config

Co-authored-by: Harshil2107 <harshilp2107@gmail.com>
Co-authored-by: kunpai <kunpai@ucdavis.edu>

helloparthshah committed:
Formatted to work with the linter

Co-authored-by: Harshil2107 <harshilp2107@gmail.com>
Co-authored-by: kunpai <kunpai@ucdavis.edu>

helloparthshah committed:
Resolved conflicts

Co-authored-by: Harshil2107 <harshilp2107@gmail.com>
Co-authored-by: kunpai <kunpai@ucdavis.edu>

Harshil2107 committed:
Fixed tests for obtain resources

Co-authored-by: Parth Shah <helloparthshah@gmail.com>
Co-authored-by: kunpai <kunpai@ucdavis.edu>

helloparthshah committed:
FIxes with tests

Co-authored-by: Harshil2107 <harshilp2107@gmail.com>
Co-authored-by: kunpai <kunpai@ucdavis.edu>

helloparthshah committed:
Updated workloads to work with unwrapped fields

Co-authored-by: kunpai <kunpai@ucdavis.edu>

helloparthshah committed:
Wrote tests for the wrapper

Co-authored-by: kunpai <kunpai@ucdavis.edu>
Co-authored-by: Harshil2107 <harshilp2107@gmail.com>

kunpai committed:
Made tests work
Fixed imports
Added to Sconscript

Co-authored-by: Parth Shah <helloparthshah@gmail.com>
Co-authored-by: Hashirubhai <harshilp2107@gmail.com>

kunpai committed:
Changes to make tests work
MongoClient now uses requests instead with read-only token
(requests is to circumvent import issues with pymongo)

Co-authored-by: Parth Shah <helloparthshah@gmail.com>

kunpai committed:
- Fixed tests
- Changed resource.py to match schema implementation
- Reverted run_pyunit to original ver
- Added resource_version everywhere
- Added optional parameters to places where necessary

Co-authored-by: Parth Shah <helloparthshah@gmail.com>

kunpai committed:
Finished Python wrapper
All tests pass
cleaned up code worked on feedback

Co-authored-by: Harshil2107 <harshilp2107@gmail.com>
Co-authored-by: Parth Shah <helloparthshah@gmail.com>

helloparthshah committed:
Reformatted using black

Co-authored-by: kunpai <kunpai@ucdavis.edu>
Co-authored-by: Harshil2107 <harshilp2107@gmail.com>

kunpai committed:
Changes to make workload tests and workloads work

Co-authored-by: Harshil2107 <harshilp2107@gmail.com>
Co-authored-by: Parth Shah <helloparthshah@gmail.com>

helloparthshah committed:
Reformatted using black

kunpai committed:
Changed disk_image to diskimage to match Schema

Co-authored-by: Harshil2107 <harshilp2107@gmail.com>
Co-authored-by: Parth Shah <helloparthshah@gmail.com>

powerjg commented on configs/gem5Resources.config.json:
Shouldn't go in configs. A gem5 "config" is different than this file.

kunpai replied:
Where is a good place we can place it?

powerjg replied:
The root would be a better location. However, Bobby and I need to chat about how to resolve that location from the stdlib. I don't like the `__file_path__.parent.parent.parent...` thing.

powerjg commented on src/python/gem5/components/boards/kernel_disk_workload.py:
```diff
@@ -94,30 +94,30 @@ def get_disk_device(self) -> str:
 
     @abstractmethod
-    def _add_disk_to_board(self, disk_image: DiskImageResource) -> None:
+    def _add_disk_to_board(self, diskimage: DiskImageResource) -> None:
```
Why did you change this variable name? Please revert unless absolutely necessary.

powerjg commented on src/python/gem5/resources/api/client.py:
```diff
@@ -0,0 +1,105 @@
+            if resource["resource_version"] == resource_version:
+                if core.gem5Version not in resource["gem5_versions"]:
+                    warnings.warn(
+                        f"Resource compatible with gem5 version: '{core.gem5Version}' not found.\n"
```
Line length.

powerjg commented on src/python/gem5/resources/api/client.py:
```diff
@@ -0,0 +1,105 @@
+        :raises: RuntimeError if the resource with the specified version is not found.
+        """
+        for resource in resources:
+            if resource["resource_version"] == resource_version:
```
```suggestion
resource = next([resource for resource in resources if resource["resource_version"] == resource_version), None)
if not resource:
    raise ...
if core.gem5Version not in resource["gem5_versions"]:
    ...
```

I don't love the nestedness.

powerjg commented on src/python/gem5/resources/api/client.py:
```diff
@@ -0,0 +1,105 @@
+        If no compatible resources are found, the original list of resources
+        is returned.
+        """
+        compatible_resources = []
+        for resource in resources:
+            if core.gem5Version in resource["gem5_versions"]:
+                compatible_resources.append(resource)
```
```suggestion
        compatible_resources = [resource for resource in resources if core.gem5Version in resource["gem5_versions"]]
```

powerjg commented on src/python/gem5/resources/api/client.py:
```diff
@@ -0,0 +1,105 @@
+                "Resource versions can be found at: "
+                f"https://gem5vision.github.io/gem5-resources-website/resources/{resources[0]['id']}/versions"
+            )
+            return resources
```
Add a comment about why you are returning *all* the resources here.

powerjg commented on src/python/gem5/resources/api/client_wrapper.py:
```diff
@@ -0,0 +1,68 @@
+config = json.load(
+    open(
+        Path(__file__).parent.parent.parent.parent.parent.parent.parent
+        / "configs/gem5Resources.config.json",
```
Do you mean for this filename to be hardcoded? If so, can you add more comments? Otherwise, it should be a variable (maybe an environment variable?)

Also, I don't think that we should use this file's location for hardcoding this config. @BobbyRBruce and I can think about the better way to do this.

powerjg commented on src/python/gem5/resources/api/client_wrapper.py:
```diff
@@ -0,0 +1,68 @@
+
+def create_clients(
+    config: dict,
+):
```
Needs some documentation

powerjg commented on src/python/gem5/resources/api/client_wrapper.py:
```diff
@@ -0,0 +1,68 @@
+    )
+)
+
+clients = create_clients(config)
```
I'm a bit uncomfortable with this code running on import.

powerjg commented on configs/example/gem5_library/riscvmatched-hello.py:
This should be a separate change.

powerjg commented on src/python/gem5/resources/api/jsonclient.py:
```diff
@@ -0,0 +1,81 @@
+
+    def get_resource_obj(
+        self, resource_id: str, resource_version: Optional[str] = None
+    ) -> Dict[str, Union[str, Dict[str, str], List[str]]]:
```
I'm not sure the return type is correct here, but it's not a big deal. Maybe simply `Dict[str, Any]`

powerjg commented on src/python/gem5/resources/api/jsonclient.py:
```diff
@@ -0,0 +1,81 @@
+        If not given, the latest version compatible with current gem5 version is returned.
+        :return: The Resource as a Python dictionary.
+        """
+        # getting all the resources with the given id from MongoDB
```
Typo with MongoDB here.

powerjg commented on src/python/gem5/resources/api/mongoclient.py:
```diff
@@ -0,0 +1,94 @@
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from .client import AbstractClient
+import requests
```
As far as I can tell, this is *not* in python's standard library. See https://requests.readthedocs.io/en/latest/user/install/#install

That said... you can probably replace it with something from the stdlib. https://docs.python.org/3/library/urllib.request.html#module-urllib.request

powerjg commented on src/python/gem5/resources/api/mongoclient.py:
```diff
@@ -0,0 +1,94 @@
+            raise Exception(f"Resource with ID '{resource_id}' not found.")
+        # sorting the resources by version
+        resources.sort(
+            key=lambda x: list(map(int, x["resource_version"].split("."))),
```
You could also use this to sort by semantic versions. I don't have a strong preference, though.

https://stackoverflow.com/questions/2574080/sorting-a-list-of-dot-separated-numbers-like-software-versions

powerjg commented on src/python/gem5/resources/api/client_wrapper.py:
```diff
@@ -0,0 +1,68 @@
+def get_resource_obj(
+    resource_id,
+    resource_version: Optional[str] = None,
+    database: Optional[str] = None,
```
I'm confused by this function's name. It's the same as `AbstractClient`, but it has a different signature. 

I suggest not having the same function names with different signatures.

powerjg commented on src/python/gem5/resources/downloader.py:
```diff
@@ -523,7 +524,6 @@ def get_resource(
             with tarfile.open(download_dest) as f:
 
                 def is_within_directory(directory, target):
-
```
Don't arbitrarily remove whitespace. (I'm surprised that black wasn't run)

powerjg commented on src/python/gem5/resources/looppoint.py:
```diff
@@ -491,7 +491,6 @@ def __init__(
         with open(_path) as file:
             json_contents = json.load(file)
             for rid in json_contents:
-
```
No whitespace changes.

powerjg commented on tests/pyunit/stdlib/resources/pyunit_md5_utils_check.py:
```diff
@@ -75,7 +75,6 @@ class MD5DirTestSuite(unittest.TestCase):
     """Test cases for gem5.resources.md5_utils.md5_dir()"""
 
     def _create_temp_directory(self) -> Path:
-
```
whitespace....

powerjg commented on src/python/gem5/resources/resource.py:
```diff
@@ -27,6 +27,8 @@
 import os
 from pathlib import Path
+
+# from m5.util import warn, fatal
```
Don't comment things out. Delete them or keep them.

powerjg commented on src/python/gem5/resources/resource.py:
```diff
@@ -522,9 +555,11 @@ def _get_weights_and_simpoints_from_file(
     resource_directory: Optional[str] = None,
     download_md5_mismatch: bool = True,
+    resource_version: Optional[str] = None,
+    database: Optional[str] = None,
```
Needs to be documented in the docstring.

powerjg commented on src/python/gem5/resources/resource.py:
```diff
@@ -547,12 +582,13 @@ def obtain_resource(
 
     # Obtain the JSON resource entry for this resource
-    resource_json = get_resources_json_obj(resource_name)
+    resource_json = get_resource_obj(
```
Should this now be something like `resource_data` since it's just a dictionary?

powerjg commented on .gitignore:
Should be in a separate change.

powerjg reviewed (commented):
Overall, this looks really good. @BobbyRBruce, I tagged you in one place. Otherwise, I think it would be best for you to spend your time looking at the testing code. I didn't look at the tests at all.

There's one "high-level" thing I don't like, though I guess I could be convinced it's ok. I don't like that you changed `disk_image` to `diskimage`. This seems like an arbitrary change that just adds to the diff and doesn't actually improve anything. I would prefer to keep the name `disk_image`.

Also, look through the diff and remove any whitespace-only changes.

Finally, what is the backwards compatibility story? Do we need to have a DB running or an entirely new json file ready before merging?

powerjg reviewed (changes requested):
Overall, this looks really good. @BobbyRBruce, I tagged you in one place. Otherwise, I think it would be best for you to spend your time looking at the testing code. I didn't look at the tests at all.

There's one "high-level" thing I don't like, though I guess I could be convinced it's ok. I don't like that you changed `disk_image` to `diskimage`. This seems like an arbitrary change that just adds to the diff and doesn't actually improve anything. I would prefer to keep the name `disk_image`.

Also, look through the diff and remove any whitespace-only changes.

Finally, what is the backwards compatibility story? Do we need to have a DB running or an entirely new json file ready before merging?

kunpai committed:
Feedback related Changes
Changed diskimage to disk_image
Refactoring client code
Reset changes to RISCVMatched

Co-authored-by: Harshil2107 <harshilp2107@gmail.com>
Co-authored-by: Parth Shah <helloparthshah@gmail.com>

kunpai committed:
Restore whitespace in one file

Co-authored-by: Harshil2107 <harshilp2107@gmail.com>
Co-authored-by: Parth Shah <helloparthshah@gmail.com>

helloparthshah committed:
Fixed a few small bugs such as:
- renamed disk_image in the comments
- Reformatted using black

helloparthshah committed:
Fixed whitespace issue

BobbyRBruce commented on .gitignore:
```diff
@@ -32,3 +32,5 @@ configs/example/memcheck.cfg
 configs/dram/lowp_sweep.cfg
 .pyenv
 .vscode
+
```
Is there a good reason to have an empty line here?

BobbyRBruce commented on configs/gem5Resources.config.json:
```diff
@@ -0,0 +1,14 @@
+            "isMongo": true
+        }
+    }
+}
```
This is petty in the extreme but it's pretty standard to have a newline character at the end of a file. See here for why: https://unix.stackexchange.com/questions/18743/whats-the-point-in-adding-a-new-line-to-the-end-of-a-file

BobbyRBruce commented on src/python/gem5/resources/api/client.py:
```diff
@@ -0,0 +1,132 @@
+            warnings.warn(
+                "Resource compatible with gem5 version: "
+                f"'{core.gem5Version}' not found.\n"
+                "Resource versions can be found at: "
```
Expand this warning a little bit. As is written it doesn't state what this means for your simulation run.

I'd say: "Resource {resource['id']} version {resource_version} is not known to be compatible with gem5 version {core.gem5Version}. This may cause problems with your simulation. This resource's compatibility with different gem5 versions can be found here: ..."

This makes it clear that this might be dangerous but the simulator isn't going to stop you.

BobbyRBruce commented on src/python/gem5/resources/api/client.py:
```diff
@@ -0,0 +1,132 @@
+            )
+        return resource
+
+    def _get_compatible_resources(self, resources: List) -> List:
```
I'd change this function to return an empty list if no compatible resources are found in the list. I know why you're doing this but I feel like this function should be simple and correspond to its name.

For example, if all the resources in the list are compatible with the current version of gem5, then this function's return is identical to the case where none of them are. We're losing information.

BobbyRBruce commented on src/python/gem5/resources/api/client.py:
```diff
@@ -0,0 +1,132 @@
+
+    def _get_compatible_resources(self, resources: List) -> List:
+        """
+        Returns a list of compatible resources with the given gem5 version.
```
We're phrasing here "given gem5 version". I'd rephrase to "current gem5 version".

BobbyRBruce commented on src/python/gem5/resources/api/client.py:
```diff
@@ -0,0 +1,132 @@
+        except:
+            return False
+
+    def _search_version(self, resources: List, resource_version: str) -> dict:
```
This function should be merged into `_get_compatible_resources`. Perhaps this function can be:

```
def _get_resources_comatible(resources: : List[dict], resource_version: str =core.gem5Version ) -> List[dict]:
```

It will filter the list to those only with a compatible version.

BobbyRBruce commented on src/python/gem5/resources/api/client.py:
```diff
@@ -0,0 +1,132 @@
+        )
+        if resource is None:
+            raise Exception(
+                f"Resource {resources[0]['id']} with version '{resource_version}'"
```
I think these warnings/exceptions should be put elsewhere. In this function, as it it written, we cannot assume all the resources passed have the same id. This function could, theoretically, be used to filter all the resources we have to the list of those with known compatibility (this may be useful at some point).

I would probably opt for putting these in a separate function with a signature like:

```
check_resource_version_capatibility(resource: dict, gem5_version: str = core.gem5Version) ->None
```

You could call it when you want to run these checks.

BobbyRBruce commented on src/python/gem5/resources/api/client.py:
```diff
@@ -0,0 +1,132 @@
+            return resources
+        return compatible_resources
+
+    def _sort_resources_by_version(self, resources: List) -> List:
```
In keeping with my other comments, this signature here does not restrict a user from inputting a list of resources with different IDs. I'd change this function to be `_sort_resources`, which sorts resources based first on id, then on version.

BobbyRBruce commented on src/python/gem5/resources/api/client_wrapper.py:
```diff
@@ -0,0 +1,87 @@
+    Returns: clients: dictionary of clients for each database
+    """
+    clients = {}
+    for resource in config["resources"]:
```
Is it too late to call call this something like "clients" instead of "resources". Needless to say, the word "resources" here is getting overused.

BobbyRBruce commented on src/python/gem5/resources/api/client_wrapper.py:
```diff
@@ -0,0 +1,87 @@
+
+def create_clients(
+    config: dict,
+):
```
Add return type here 'Dict'

BobbyRBruce commented on src/python/gem5/resources/api/client_wrapper.py:
```diff
@@ -0,0 +1,87 @@
+    :return: resource object
+    """
+    if not database:
+        database = list(clients.keys())[0]
```
I'm not sure if we have discussed this before, but is this the behavior we want? If the user doesn't specify the database here, we just assume we're looking in the first one.

How about this: we don't have a "database" field, and we always search all our clients for instances of a resource. We throw an exception if there are duplicates (same id and version). Would that work?

BobbyRBruce commented on src/python/gem5/resources/api/jsonclient.py:
```diff
@@ -0,0 +1,78 @@
+
+import json
+from pathlib import Path
+import requests
```
As Jason said in the other file, I don't think we can use requests.

BobbyRBruce commented on src/python/gem5/resources/api/jsonclient.py:
```diff
@@ -0,0 +1,78 @@
+        else:
+            self.resources = requests.get(self.path).json()
+
+    def get_resource_obj_from_client(
```
It seems to me like a simpler way of doing this would be to have an abstract function in `AbstractClient` called `get_resources_by_id(id: str)`. Then have `get_resource_obj_from_client` implemented in `AbstractClient` (which calls `get_resources_by_id`).

This would mean `JSONClient` and `MongoClient` would only need to implement `get_resources_by_id` which seems to be the only thing that truly differs between the two clients.

BobbyRBruce commented on src/python/gem5/resources/downloader.py:
```diff
@@ -109,7 +110,6 @@ def _get_resources_json_at_path(path: str, use_caching: bool = True) -> Dict:
     # its Truncated Exponential Backoff algorithm
     # (maximum of roughly 1 minute). Typically this code will run quickly.
     with FileLock(f"{download_path}.lock", timeout=120):
-
```
You've deleted a line here for no real reason. Avoid doing stuff like this. It just makes the change a bit messier than it needs to be.

BobbyRBruce commented on src/python/gem5/resources/downloader.py:
```diff
@@ -41,6 +41,7 @@
 from urllib.error import HTTPError
 from typing import List, Dict, Set, Optional
 
+from python.gem5.resources.api.client_wrapper import get_resource_obj
```
Within the stdlib you should use relative imports. So in this case, `from .api.client_wrapper import get_resource_obj`

BobbyRBruce commented on src/python/gem5/resources/looppoint.py:
```diff
@@ -491,7 +491,7 @@ def __init__(
             json_contents = json.load(file)
             for rid in json_contents:
-
+                
```
This to me shows you probably don't have our pre-commit checks installed. It will automatically detect and remove unnecessary whitespace like this. Run `./util/pre-commit-install.sh` to do so

BobbyRBruce commented on src/python/gem5/resources/workload.py:
```diff
@@ -217,5 +225,4 @@ def __init__(
             for key in workload_json["additional_params"]:
                 assert isinstance(key, str)
                 params[key] = workload_json["additional_params"][key]
-
```
I'll only say this once more, because there's numerous parts this happens, but please don't remove lines randomly like this. The commit/pull-request should only be focused on adding/removing code as needed.

BobbyRBruce commented on src/python/SConscript:
```diff
@@ -268,6 +268,10 @@ PySource('gem5.resources', 'gem5/resources/resource.py')
 PySource('gem5.resources', 'gem5/resources/workload.py')
 PySource('gem5.resources', 'gem5/resources/looppoint.py')
 PySource('gem5.resources', 'gem5/resources/elfie.py')
+PySource('gem5.resources.api', 'gem5/resources/api/client_wrapper.py')
```
I think "api" is too generic. Can we call this directory "client_api" instead?

BobbyRBruce commented on src/python/gem5/resources/downloader.py:
```diff
@@ -430,17 +432,17 @@ def get_resource(
     with FileLock(f"{to_path}.lock", timeout=900):
-
-        resource_json = get_resources_json_obj(resource_name)
+        resource_obj = get_resource_obj(
```
I'm getting a bit confused by this: is this returning a resource object (i.e. a full implementation such as `BinaryResource`, or a dictionary representing the JSON?). It's really not clear.

From what I can see in the rest of this file, it's just a JSON representation of the object, so why the change of this function name and variable. This seems pointless.

BobbyRBruce commented on src/python/gem5/resources/api/client_wrapper.py:
```diff
@@ -0,0 +1,87 @@
+    config = json.load(
+        open(
+            Path(__file__).parent.parent.parent.parent.parent.parent.parent
+            / "configs/gem5Resources.config.json",
```
There is an issue here of where the config file should exist, how it should be used, and how a user interacts with it. I discussed this with Jason and I think you should do the following:

1. The config file should be called "gem5-config.json".
2. It should be moved to "src/python".
3. This configs file should be compiled into the binary (me and Jason's answer to the question "but how do I do this?" is "🤷‍♂️" , but we _think_ it should be possible...)
4. This compiled-in config file config file will serve as the default values for those fields. In addition to this the user can set an environment variable to specify their own "gem5-config.json" file to override this. (e.g. "GEM5_CONFIG="/home/bbruce/my-config.json" ./build/ALL/gem5.opt ...` should allow for loading of the "my-config.json" file, replacing the compiled in version).
5. If there is a "gem5-config.json" in the current working directory, gem5 should use this.

4 and 5 should be optional stretch-goals for now (i.e., only do them if you've got nothing better to do and want to). but I think you should try to do 1, 2, and 3.

BobbyRBruce reviewed (commented):
I haven't reviewed the tests (i'll do so another time).

Even though this looks like a lot of criticism, this is really excellent work. Just a couple more iterations of this and I think we'll be ready to incorporate into the gem5 project.

kunpai committed:
resources: Worked on client feedback

Change-Id: I64bdb84b03cd552b7e7852c388d1269105f8a57d

kunpai committed:
resources: Merging

Change-Id: I90fa155dea669f096fb7fafcebfcc90d203c9e10
Co-authored-by: Harshil2107 <harshilp2107@gmail.com>
Co-authored-by: Parth Shah <helloparthshah@gmail.com>

kunpai committed:
resources: Worked on client feedback

Mocked urllib in tests
Added multidatabase support
Worked on other minor feedback

Change-Id: I069ffc4ff390ffbb4ff22a951a2e44c2da75e31d
Co-authored-by: Parth Shah <helloparthshah@gmail.com>
Co-authored-by: Harshil2107 <harshilp2107@gmail.com>

kunpai committed:
resources: gem5-config changes

Can copy into build directory
Can use env variables and overriding it is possible
Error messages for the same

Change-Id: I420fa81ef49a6bd661f182751f2725c2c1f522fb
Co-authored-by: Parth Shah <helloparthshah@gmail.com>
Co-authored-by: Harshil2107 <harshilp2107@gmail.com>

kunpai committed:
tests: Added tests for multiple databases

Change-Id: I7e90a0bc626ed6f50338b62cc007aa823f38823f
Co-authored-by: Harshil2107 <harshilp2107@gmail.com>
Co-authored-by: Parth Shah <helloparthshah@gmail.com>
